packages/circuit-ui/components/ToastContext/ToastContext.tsx

```diff
@@ -0,0 +1,158 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  createContext,
+  ReactNode,
+  useCallback,
+  useEffect,
+  useMemo,
+} from 'react';
+import { useClickTrigger } from '@sumup/collector';
+import { css } from '@emotion/react';
+
+import { useStack, StackItem } from '../../hooks/useStack';
+import styled, { StyleProps } from '../../styles/styled';
+import { spacing } from '../..';
+
+import { BaseToastProps, ToastComponent } from './types';
+
+type ToastState<TProps extends BaseToastProps> = TProps &
+  StackItem & { component: ToastComponent<TProps> };
+
+type ToastContextValue = {
+  /* eslint-disable @typescript-eslint/no-explicit-any */
+  setToast: (toast: ToastState<any>) => void;
+  removeToast: (toast: ToastState<any>) => void;
+  /* eslint-enable @typescript-eslint/no-explicit-any */
+};
+
+export const ToastContext = createContext<ToastContextValue>({
+  setToast: () => {},
+  removeToast: () => {},
+});
+
+export interface ToastProviderProps {
+  /**
+   * The ToastProvider should wrap your entire application.
+   */
+  children: ReactNode;
+}
+
+const listWrapperStyles = ({ theme }: StyleProps) => css`
+  position: fixed;
+  bottom: ${theme.spacings.giga};
+  left: 50%;
+  transform: translateX(-50%);
+  display: flex;
+  flex-direction: column-reverse;
+`;
+
+const ListWrapper = styled('ul')(listWrapperStyles);
+
+export function ToastProvider<TProps extends BaseToastProps>({
+  children,
+}: ToastProviderProps): JSX.Element {
+  const sendEvent = useClickTrigger();
+  const [toasts, dispatch] = useStack<ToastState<TProps>>([]);
+
+  const setToast = useCallback(
+    (toast: ToastState<TProps>) => {
+      if (toast.tracking && toast.tracking.label) {
+        sendEvent({
+          component: 'toast',
+          ...toast.tracking,
+          label: `${toast.tracking.label}|open`,
+        });
+      }
+
+      dispatch({
+        type: 'push',
+        item: toast,
+      });
+    },
+    [dispatch, sendEvent],
+  );
+
+  const removeToast = useCallback(
+    (toast: ToastState<TProps>) => {
+      if (toast.tracking && toast.tracking.label) {
+        sendEvent({
+          component: 'toast',
+          ...toast.tracking,
+          label: `${toast.tracking.label}|close`,
+        });
+      }
+      if (toast.onClose) {
+        toast.onClose();
+      }
+      dispatch({
+        type: 'remove',
+        id: toast.id,
+        timeout: toast.component.TIMEOUT,
+      });
+    },
+    [dispatch, sendEvent],
+  );
+
+  const context = useMemo(() => ({ setToast, removeToast }), [
+    setToast,
+    removeToast,
+  ]);
+
+  useEffect(() => {
+    const toastToDismiss = toasts[0];
+    if (!toastToDismiss) {
+      return undefined;
+    }
+    const duration = toastToDismiss.duration || 3000;
```

Comment:
This would mean that **all** toasts are auto-dismissed, even if the developer doesn't specify a `duration`. Is this the desired behavior? I assumed that a toast would stay indefinitely if no duration is specified. 

Either way, 3000ms is (also) the _minimum_ duration:

```suggestion
    const duration = toastToDismiss.duration ? Math.max(toastToDismiss.duration, 3000) : 3000;
```

Reply:
Yes, the toast is always auto-dismissed.

Sighted keyboard users need them to disappear, because they take up screen space and users shouldn't have to go through the DOM to find them and dismiss them manually

(but yes, it should be minimum 3000ms—or arguably even more—, that's right)

Reply:
I figured that toasts that don't auto dismiss would always have the close button. Good point about the keyboard accessibility though 👍🏻